### PR TITLE
fix submeshes have different instancing options

### PR DESCRIPTION
**Pull request by jk20012001**
Re: #

### Changelog

*

-------

### Continuous Integration

This pull request:

* [x] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

**Comment by troublemaker52025** [comment]
@cocos-robot run test cases

File: cocos/core/pipeline/custom/executor.ts [hunk 1]
```diff
@@ -801,30 +801,30 @@ class DevicePreSceneTask extends WebSceneTask {
...
+                const passes = subModel.passes;
                 for (const p of passes) {
                     if (p.phase !== this._currentQueue.phaseID) continue;
                     const batchingScheme = p.batchingScheme;
                     // If the size of instances is not 0, it has been added
                     if (batchingScheme === BatchingSchemes.INSTANCING
                         && !this._submitInfo.instances.size) {
                         const instancedBuffer = p.getInstancedBuffer();
-                        instancedBuffer.merge(submodel, ro.model.instancedAttributes, passes.indexOf(p));
+                        instancedBuffer.merge(subModel, ro.model.getInstancedAttributeBlock(subModel), passes.indexOf(p));
```
Comment (jk20012001) [inline cocos/core/pipeline/custom/executor.ts]: this get function is named as same as native engine

File: native/cocos/renderer/pipeline/InstancedBuffer.cpp [hunk 2]
```diff
@@ -50,15 +50,16 @@ void InstancedBuffer::destroy() {
     _instances.clear();
 }
 
-void InstancedBuffer::merge(const scene::Model *model, const scene::SubModel *subModel, uint32_t passIdx) {
-    merge(model, subModel, passIdx, nullptr);
```
Comment (jk20012001) [inline native/cocos/renderer/pipeline/InstancedBuffer.cpp:54]: arguments are modified as same as ts engine

File: cocos/core/renderer/scene/model.ts [hunk 3]
```diff
@@ -297,10 +288,20 @@ export class Model {
     public isDynamicBatching = false;
 
     /**
-     * @en The instance attributes
-     * @zh 实例化属性
+     * @en The instance attribute block, access by subModel
```
Comment (minggo) [inline cocos/core/renderer/scene/model.ts]: subModel -> sub model

File: cocos/core/renderer/scene/model.ts [hunk 4]
```diff
@@ -297,10 +288,20 @@ export class Model {
...
+     * @zh 硬件实例化属性，通过子模型访问
      */
-    public instancedAttributes: IInstancedAttributeBlock = { buffer: null!, views: [], attributes: [] };
+    public getInstancedAttributeBlock (subModel: SubModel): IInstancedAttributeBlock {
+        if (!this._instancedAttributes.has(subModel)) {
+            this._instancedAttributes.set(subModel, { buffer: null!, views: [], attributes: [] });
+        }
```
Comment (minggo) [inline cocos/core/renderer/scene/model.ts]: It is not a good idea to add an instanced attribute in get function.

File: cocos/core/renderer/scene/model.ts [hunk 5]
```diff
@@ -158,14 +157,6 @@ export class Model {
         return this._updateStamp;
     }
 
-    /**
-     * @en Whether GPU instancing is enabled for the current model
-     * @zh 是否开启实例化渲染
-     */
-    get isInstancingEnabled () {
-        return this._instMatWorldIdx >= 0;
-    }
```
Comment (minggo) [inline cocos/core/renderer/scene/model.ts:167]: Will it break compatibility.

File: cocos/core/renderer/scene/model.ts [hunk 6]
```diff
@@ -297,10 +288,20 @@ export class Model {
...
      */
-    public instancedAttributes: IInstancedAttributeBlock = { buffer: null!, views: [], attributes: [] };
+    public getInstancedAttributeBlock (subModel: SubModel): IInstancedAttributeBlock {
+        if (!this._instancedAttributes.has(subModel)) {
+            this._instancedAttributes.set(subModel, { buffer: null!, views: [], attributes: [] });
+        }
+        return this._instancedAttributes.get(subModel) as IInstancedAttributeBlock;
+    }
+
+    get instancedAttributes () {
```
Comment (minggo) [inline cocos/core/renderer/scene/model.ts]: I think should add readonly for return value.

File: cocos/core/renderer/scene/model.ts [hunk 7]
```diff
@@ -375,10 +376,10 @@ export class Model {
     protected _localBuffer: Buffer | null = null;
 
     /**
-     * @en Instance matrix id
-     * @zh 实例矩阵索引
+     * @en Instance matrix id, access by subModel
```
Comment (minggo) [inline cocos/core/renderer/scene/model.ts]: subModel -> sub model

File: cocos/core/renderer/scene/model.ts [hunk 8]
```diff
@@ -297,10 +288,20 @@ export class Model {
...
+        if (!this._instancedAttributes.has(subModel)) {
+            this._instancedAttributes.set(subModel, { buffer: null!, views: [], attributes: [] });
+        }
+        return this._instancedAttributes.get(subModel) as IInstancedAttributeBlock;
+    }
+
+    get instancedAttributes () {
+        return this._instancedAttributes;
+    }
+    protected _instancedAttributes: Map<SubModel, IInstancedAttributeBlock> = new Map<SubModel, IInstancedAttributeBlock>();
```
Comment (minggo) [inline cocos/core/renderer/scene/model.ts]: How about rename to `_instancedAttributeMap `, as `_instancedAttributes ` is always used for vector.

File: cocos/core/renderer/scene/model.ts [hunk 9]
```diff
@@ -375,10 +376,10 @@ export class Model {
     protected _localBuffer: Buffer | null = null;
 
     /**
-     * @en Instance matrix id
-     * @zh 实例矩阵索引
+     * @en Instance matrix id, access by subModel
+     * @zh 硬件实例化中的矩阵索引，通过子模型访问
      */
-    private _instMatWorldIdx = -1;
+    private _instMatWorldIdx: Map<SubModel, number> = new Map<SubModel, number>();
```
Comment (minggo) [inline cocos/core/renderer/scene/model.ts]: Not it is a map, so i think should rename the variable too, such as `_instancedMatrixWorldIndexMap`.

File: cocos/3d/models/baked-skinning-model.ts [hunk 10]
```diff
@@ -204,20 +213,24 @@ export class BakedSkinningModel extends MorphModel {
...
-        this._instAnimInfoIdx = this._getInstancedAttributeIndex(INST_JOINT_ANIM_INFO);
+    protected _updateInstancedAttributes (attributes: Attribute[], subModel: SubModel) {
+        super._updateInstancedAttributes(attributes, subModel);
+        this._instAnimInfoIdx = this._getInstancedAttributeIndex(subModel, INST_JOINT_ANIM_INFO);
         this.updateInstancedJointTextureInfo();
     }
 
     private updateInstancedJointTextureInfo () {
         const { jointTextureInfo, animInfo } = this._jointsMedium;
+        const values = Array.from(this._instancedAttributes.values());
```
Comment (star-e) [inline cocos/3d/models/baked-skinning-model.ts]: Will `updateInstancedJointTextureInfo` be called frequently?

File: cocos/core/renderer/scene/model.ts (hunk 5 above)
Comment (jk20012001) [reply cocos/core/renderer/scene/model.ts:167]: this function is not accuracy enough,  only used by planar-shadow-queue

**Review by stanleyljl — APPROVED** [review]

File: cocos/core/renderer/scene/model.ts (hunk 8 above)
Comment (minggo) [reply cocos/core/renderer/scene/model.ts]: I think it is better to record the attribute block and world matrix index in SubModel, then can avoid map operation.

File: cocos/3d/models/baked-skinning-model.ts [hunk 11]
```diff
@@ -129,11 +130,19 @@ export class BakedSkinningModel extends MorphModel {
...
-        const idx = this._instAnimInfoIdx;
-        if (idx >= 0) {
-            const view = this.instancedAttributes.views[idx];
-            view[0] = info.data[0];
-        } else if (info.dirty) {
+
+        let hasNonInstancingPass = false;
+        for (let i = 0; i < this._subModels.length; i++) {
+            const subModel = this._subModels[i];
+            const idx = this._instAnimInfoIdx;
```
Comment (minggo) [inline cocos/3d/models/baked-skinning-model.ts]: I am not sure if define variable outside for loop is better in ts.

File: cocos/core/renderer/scene/model.ts [hunk 12]
```diff
@@ -558,11 +532,18 @@ export class Model {
...
-            const attrs = this.instancedAttributes.views;
-            uploadMat4AsVec4x3(worldMatrix, attrs[idx], attrs[idx + 1], attrs[idx + 2]);
-        } else if (this._localBuffer) {
+        let hasNonInstancingPass = false;
+        for (let i = 0; i < subModels.length; i++) {
+            const subModel = subModels[i];
+            const idx = subModel.instancedWorldMatrixIndex;
+            if (idx >= 0) {
+                const attrs = subModel.instancedAttributeBlock.views;
+                uploadMat4AsVec4x3(worldMatrix, attrs[idx], attrs[idx + 1], attrs[idx + 2]);
```
Comment (minggo) [inline cocos/core/renderer/scene/model.ts]: It is not good to modify subModel.instancedAttributeBlock outside of SubModel.

File: cocos/core/renderer/scene/submodel.ts [hunk 13]
```diff
@@ -175,6 +183,25 @@ export class SubModel {
         return this._planarShader;
     }
 
+    /**
+     * @en The instance attribute block, access by sub model
+     * @zh 硬件实例化属性，通过子模型访问
+     */
+    get instancedAttributeBlock () {
```
Comment (minggo) [inline cocos/core/renderer/scene/submodel.ts:191]: Should add readonly for it. And can not modify it outside of this class.

File: native/cocos/renderer/pipeline/PlanarShadowQueue.cpp [hunk 14]
```diff
@@ -88,15 +88,21 @@ void PlanarShadowQueue::gatherShadowPasses(scene::Camera *camera, gfx::CommandBu
...
 
-        if (!model->getInstanceAttributes().empty()) {
-            int i = 0;
-            for (const auto &subModel : model->getSubModels()) {
-                instancedBuffer->merge(model, subModel, i, subModel->getPlanarInstanceShader());
-                _instancedQueue->add(instancedBuffer);
-                ++i;
+        const auto &subModels = model->getSubModels();
+        const auto subModelCount = subModels.size();
+        for (uint32_t subModelIdx = 0; subModelIdx < subModelCount; ++subModelIdx) {
```
Comment (minggo) [inline native/cocos/renderer/pipeline/PlanarShadowQueue.cpp]: Can use for in like
```c++
for (const auto &subModel : subModels) {
}
```

File: native/cocos/renderer/pipeline/PlanarShadowQueue.cpp [hunk 15]
```diff
@@ -88,15 +88,21 @@ void PlanarShadowQueue::gatherShadowPasses(scene::Camera *camera, gfx::CommandBu
...
+            const auto &subModel = subModels[subModelIdx];
+            const auto &subModelPasses = subModel->getPasses();
+            for (index_t i = 0; i < static_cast<index_t>(subModelPasses.size()); ++i) {
```
Comment (minggo) [inline native/cocos/renderer/pipeline/PlanarShadowQueue.cpp:94]: The same as above.

File: native/cocos/renderer/pipeline/deferred/GbufferStage.cpp [hunk 16]
```diff
@@ -99,17 +100,17 @@ void GbufferStage::dispenseRenderObject2Queues() {
     for (auto ro : renderObjects) {
         const auto *const model = ro.model;
         const auto &subModels = model->getSubModels();
-        auto subModelCount = subModels.size();
+        const auto subModelCount = subModels.size();
         for (subModelIdx = 0; subModelIdx < subModelCount; ++subModelIdx) {
```
Comment (minggo) [inline native/cocos/renderer/pipeline/deferred/GbufferStage.cpp:104]: The same as above.

File: native/cocos/renderer/pipeline/PlanarShadowQueue.cpp (hunk 15 above)
Comment (jk20012001) [reply native/cocos/renderer/pipeline/PlanarShadowQueue.cpp:94]: merge function needs pass index so don't need change here

File: native/cocos/renderer/pipeline/deferred/GbufferStage.cpp (hunk 16 above)
Comment (jk20012001) [reply native/cocos/renderer/pipeline/deferred/GbufferStage.cpp:104]: the same as above

File: cocos/core/renderer/scene/submodel.ts [hunk 17]
```diff
@@ -375,6 +403,89 @@ export class SubModel {
...
+    /**
+     * @en
+     * update instancing related data, invoked by model
+     * @zh
+     * 更新硬件实例化相关数据，一般由model调用
+     */
+    /**
+     * @internal
+     */
+    public UpdateInstancedAttributes (attributes: Attribute[]) {
```
Comment (dumganhar) [inline cocos/core/renderer/scene/submodel.ts:450]: Don't use capital letter as prefix.

**Pull request merged**